### PR TITLE
Configurable upstream manager module

### PR DESCRIPTION
**Pull request by Tieske**
the options table can now contain an optional option named `upstream_mgr`. If not provided, it will default to the normal resty upstream module. If given it must either be a module name to load, or the module table itself. see issue #11 

Obviously, the interface of the module provided must be identical to the default resty upstream module.

Note: the `status_page` function now takes an options table, but with only one option. The intent is to enable future changes without breaking backward compatibility (eg. instead of a page for all upstreams, specify a single one, or setting a flag to return a table or json with statusses instead of a page)

### CODE REVIEW

**Comment by agentzh** [comment]
@Tieske Maybe the option `manager_module` is a better name than `upstream_mgr`?

File: lib/resty/upstream/healthcheck.lua [hunk 1]
```diff
@@ -490,6 +480,28 @@ check = function (premature, ctx)
     end
 end
 
+local function load_upstream_mgr(upstream_mgr)
+    upstream_mgr = upstream_mgr or "ngx.upstreams"
+
+    assert(type(upstream_mgr) == "string" or type(upstream_mgr == "table"),
+           "the upstream_mgr option must be a string (module name) or table (actual module)")
```
Comment (doujiang24) [inline lib/resty/upstream/healthcheck.lua]: this line exceed 80 chars, maybe you can use [lua-releng](https://github.com/openresty/nginx-devel-utils/blob/master/lua-releng)

**Comment by Tieske** [comment]
Maybe full name `upstream_manager`? I don't like `manager_module` because they are two words both not saying anything specific (might as well be a component of a taco dispenser in a supermarket)

**Comment by agentzh** [comment]
@Tieske LOL. `upstream_manager` then.

**Comment by Tieske** [comment]
On second thought, we might need some more updates to reach the required level of flexibility. Also had a comment from @doujiang24 on that.

Some additional requirements;

 - (incoming) signal for the healthchecker to reload the primairy and backup peer data (when we add or remove an upstream peer)
 - (outgoing) signal when there is a change in _available_ upstream peers. Callback maybe, including the list of active peers. A loadbalancer should hook into this.

Thoughts? 
Should we be adding those elements in this PR or separate?

**Comment by agentzh** [comment]
@Tieske Feel free to add stuff to this PR if you want to :)

**Comment by agentzh** [comment]
@Tieske You're currently the only person using this feature anyway :)

File: lib/resty/upstream/healthcheck.lua [hunk 2]
```diff
@@ -490,6 +480,28 @@ check = function (premature, ctx)
     end
 end
 
+local function load_upstream_mgr(upstream_mgr)
+    upstream_mgr = upstream_mgr or "ngx.upstreams"
+
+    assert(type(upstream_mgr) == "string" or type(upstream_mgr == "table"),
```
Comment (vesafb) [inline lib/resty/upstream/healthcheck.lua]: should be type(upstream_mgr) == "table"?

**Comment by Tieske** [comment]
we've diverted into other ways which no longer use the upstreams defined in the config. Hence closing this.

**Pull request closed without merging**